### ISSUE TITLE
Make test_updateWithKeywords work on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6082]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6082|
|Type|enhancement|
|Created|2012-10-08 12:55:24Z|
|Branch|https://github.com/twisted/twisted/tree/http-headers-update-keywords-6082|

twisted.web.test.test_http_headers._DictHeader.update works like dict.update, accepting keyword arguments.  Since header names should be bytes, though, some extra support is needed to make it correct on Python 3.

Skipped porting this functionality as part of [#6080](https://github.com/twisted/twisted/issues/6080) since it's somewhat obscure and slightly more work, but eventually it should be supported.

Attachments:

* [headers-keyword-update-6082-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/798/798e88b671c36396b6bd4ee08dd8e0fd7ee6de62/261e84b75cef8b6fd947c2c96c2e2b4bb7ef0196.patch) (1229 bytes) - added by Unit03 on 2015-11-16 16:49:25Z - 
* [headers-keyword-update-6082-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/798/798e88b671c36396b6bd4ee08dd8e0fd7ee6de62/92940ccd7b364fa1e067f301c706a53e04ce2319.patch) (1610 bytes) - added by Unit03 on 2015-11-18 20:42:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6082 6082
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_http_headers_update_keywords_6082 branches/http-headers-update-keywords-6082
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1349700924000000 1349700924000000
changetime__1447917025820729 1447917025820729
version__None None
owner__Unit03 Unit03
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Unit03's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Unit03 commented|
|-|-|

This is seemingly more trivial than expected, but maybe I'm missing something.


```t.w.http_headers._DictHeaders``` stores headers with keys as bytes, but for keyword-style update of the inner dict, keyword names - which are unicode/str - must first be encoded in ```__setitem__```. I wasn't sure about which encoding to use.

Does the patch look reasonable?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

(In [46240]) Branching to http-headers-update-keywords-6082.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to Unit03|
|-|-|

Many thanks for your patch!

Is 'ascii' the right encoding for headers?
From what I understand from RFC 2616 I think that latin1 characters are also allowed as part of the headers.

Looking forward for your comments.

I have applied your latest patch and sent it to our builbot.

In case latin1 is not supported, please add an explicit test and update the documentation to let user know that even if RFC 2616 allows latin1, Twisted does not yet support that.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="Unit03's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Unit03 set owner to @adiroiban|
|-|-|

Then I guess it should be ```latin1```. But

  In case latin1 is not supported (...)

Not supported where?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Not supported by Twisted.

--------

I think that ASCII is fine. I was confused ... but now I think that latin-1 was only supported as header value, not as header name

​https://tools.ietf.org/html/rfc7230#section-3.2.4

   Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.

----------

So. Your patch is great.

It still needs  a news file fragment so that we will have a note in the release note and people will know that this bug is fixed.

Here is some more info about how we manage the news file:
http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

For this case, feel fee to just the release notes as a comment and I will take care of creating the right file.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to Unit03|
|-|-|

Please provide a text for the release notes associated with this patch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="Unit03's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Unit03 set owner to @adiroiban|
|-|-|

Great, thanks. :)

Sorry about lack of news file, I wrongly assumed I'm not supposed to write it. I'm attaching updated patch with news entry.

Feel free to reword it if necessary, though (I wasn't sure about few things, e.g. whether to include that t.w.http_headers is now fully ported to 3 - it wasn't mentioned in NEWS previously).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to Unit03|
|-|-|

Thanks.

Changes applied.

Will wait for test result and if all is ok will merge.

Many thanks again for your help!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46255]) Merge http-headers-update-keywords-6082: Fix updating twisted.web.http_headers._DictHeaders using keywords in Python 3.

Author: Unit03
Reviewer: adiroiban
Fixes: [#6082](https://github.com/twisted/twisted/issues/6082)
